File: productArray.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) 
    {
        int n = nums.size();
        vector<int> prefixProduct (n, 1);
        vector<int> suffixProduct (n ,1);
        vector<int> answer;
        
        prefixProduct[0] = nums[0];
        suffixProduct[n - 1] = 1;
        
        for(int i = 1; i < n; i++)
            prefixProduct[i] = prefixProduct[i - 1] * nums[i];
        
        for(int i = n - 2; i >= 0; i--)
            suffixProduct[i] = suffixProduct[i + 1] * nums[i + 1];
        
        for(int i = 0, left = 1, right = 1; i < n; i++)
        {
            if(i == 0)
                left = 1;
            else
                left = prefixProduct[i - 1];
            
            right = suffixProduct[i];
            
            answer.push_back(left * right);
            
        }
        return answer;
    }
};

File: progress_sheet/productArray.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        size = len(nums)
        left_product = [1] * (size + 1)
        right_product = [1] * (size + 1)
        nums.append(1)

        # Store left prefix product
        for i in range(size):
            left_product[i] = left_cproduct[i - 1] * nums[i]
        
        # Store right prefix product
        for i in range(size - 1, -1, -1):
            right_product[i] = right_product[i + 1] * nums[i]

        # multiply left product and right product excluding the current element
        for i in range(size):
            nums[i] = right_product[i + 1] * left_product[i - 1]
        
        nums.pop()

        return nums
